## <center> Mathematical Foundations </center>
## <center> MFE Spring 2021 </center>
## <center> Assignment 2 </center>
<center>Student Name: Rick Wuebker
<center>ID: richard_wuebker
<center> Solved in team: No

# 1. Stochastic Matrices
><b>Given the below transition matrix</b>

In [1]:
import numpy as np
import pandas as pd
np.set_printoptions(linewidth=100)
np.set_printoptions(suppress=True, threshold=np.inf)
pd.set_option('display.float_format', lambda x: '%.8f' % x)
l = [[0.90788, 0.08291, 0.00716, 0.00102, 0.00102, 0, 0, 0],
     [0.00103, 0.91219, 0.07851, 0.00620, 0.00103, 0.00103, 0, 0],
     [0.00924, 0.02361, 0.90041, 0.05441, 0.00719, 0.00308, 0.00103, 0.00103],
     [0, 0.00318, 0.05938, 0.86947, 0.05302, 0.011166, 0.00117, 0.00212],
     [0, 0.00110, 0.00659, 0.07692, 0.80549, 0.08791, 0.00989, 0.01209],
     [0, 0.00114, 0.00227, 0.00454, 0.06470, 0.82747, 0.04086, 0.05902],
     [0, 0, 0.00456, 0.01251, 0.02275, 0.12856, 0.60637, 0.22526],
     [0, 0, 0, 0, 0, 0, 0, 1.0]]
phi = np.array(l)
ix = ['AAA', 'AA', 'A', 'BBB', 'BB', 'B', 'CCC', 'Default']
df = pd.DataFrame(data = phi, columns=ix, index=ix)
df

AAA         AA          A        BBB         BB          B  \
AAA     0.90788000 0.08291000 0.00716000 0.00102000 0.00102000 0.00000000   
AA      0.00103000 0.91219000 0.07851000 0.00620000 0.00103000 0.00103000   
A       0.00924000 0.02361000 0.90041000 0.05441000 0.00719000 0.00308000   
BBB     0.00000000 0.00318000 0.05938000 0.86947000 0.05302000 0.01116600   
BB      0.00000000 0.00110000 0.00659000 0.07692000 0.80549000 0.08791000   
B       0.00000000 0.00114000 0.00227000 0.00454000 0.06470000 0.82747000   
CCC     0.00000000 0.00000000 0.00456000 0.01251000 0.02275000 0.12856000   
Default 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000   

               CCC    Default  
AAA     0.00000000 0.00000000  
AA      0.00000000 0.00000000  
A       0.00103000 0.00103000  
BBB     0.00117000 0.00212000  
BB      0.00989000 0.01209000  
B       0.04086000 0.05902000  
CCC     0.60637000 0.22526000  
Default 0.00000000 1.00000000

><b>(a) What is the probability that a bond over two years moves from an AA rating to Default?</b>

In [2]:
phi_sq = phi.dot(phi)
print(phi_sq[0,7])
pd.DataFrame(data=phi_sq, columns=ix, index=ix)

2.1869e-05


AAA         AA          A        BBB         BB          B  \
AAA     0.82439765 0.15107542 0.01952391 0.00279497 0.00193860 0.00020851   
AA      0.00260010 0.83405164 0.14269188 0.01540298 0.00273011 0.00219344   
A       0.01673292 0.04374602 0.81594788 0.09703490 0.01540669 0.00671822   
BBB     0.00055195 0.00713869 0.10572522 0.76337229 0.08998565 0.02394560   
BB      0.00006202 0.00238986 0.01614041 0.12972614 0.65885373 0.14570529   
B       0.00002215 0.00212241 0.00489407 0.01332257 0.10684028 0.69570620   
CCC     0.00004213 0.00031903 0.00805551 0.02104446 0.04113371 0.18648815   
Default 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000   

               CCC    Default  
AAA     0.00001866 0.00002187  
AA      0.00014039 0.00016725  
A       0.00181260 0.00257350  
BBB     0.00276850 0.00558802  
BB      0.01765208 0.02941450  
B       0.05923424 0.11785559  
CCC     0.37318187 0.36974478  
Default 0.00000000 1.00000000

$$ \Phi^2 = \begin{bmatrix}
0.82439765 & 0.15107542 & 0.01952391 & 0.00279497 &  0.0019386 &0.00020851& 0.00001866& 0.00002187 \\ 
0.0026001 & 0.83405164& 0.14269188& 0.01540298& 0.00273011& 0.00219344& 0.00014039& 0.00016725 \\
0.01673292& 0.04374602& 0.81594788& 0.0970349 & 0.01540669& 0.00671822& 0.0018126 & 0.0025735 \\
0.00055195& 0.00713869& 0.10572522& 0.76337229& 0.08998565& 0.0239456 & 0.0027685 & 0.00558802 \\
0.00006202& 0.00238986& 0.01614041& 0.12972614& 0.65885373& 0.14570529& 0.01765208& 0.0294145 \\
0.00002215& 0.00212241& 0.00489407& 0.01332257& 0.10684028& 0.6957062 & 0.05923424& 0.11785559 \\
0.00004213& 0.00031903& 0.00805551& 0.02104446& 0.04113371& 0.18648815& 0.37318187& 0.36974478 \\
0.        & 0.        & 0.        & 0.        & 0.        &0.        & 0.        & 1. \\
\end{bmatrix}$$

<center>$ \Phi^2[0,7] = 0.002187$%</center>

><b>(b) What is the long-term (stationary) distribution, $\mathbf{p}$, of the ratings for a corporate bond with the above transition matrix (assuming that it has a very long maturity state)?

In [11]:
new_phi = phi
converged = False
for i in range(100000):
    old_phi = new_phi
    new_phi = old_phi.dot(phi)
    if np.allclose(new_phi, old_phi, atol=1e-10): 
        print(f'converged after {i} iterations!!!')
        print(pd.DataFrame(new_phi, columns=ix, index=ix))
        converged = True
        break

if not converged:
    print('failed to converge')

    
    

converged after 1460 iterations!!!
               AAA         AA          A        BBB         BB          B  \
AAA     0.00000000 0.00000000 0.00000001 0.00000001 0.00000000 0.00000000   
AA      0.00000000 0.00000000 0.00000001 0.00000001 0.00000000 0.00000000   
A       0.00000000 0.00000000 0.00000001 0.00000000 0.00000000 0.00000000   
BBB     0.00000000 0.00000000 0.00000001 0.00000000 0.00000000 0.00000000   
BB      0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000   
B       0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000   
CCC     0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000   
Default 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000   

               CCC    Default  
AAA     0.00000000 0.98875433  
AA      0.00000000 0.98882510  
A       0.00000000 0.98889601  
BBB     0.00000000 0.98763573  
BB      0.00000000 0.99281930  
B       0.00000000 0.99625655  
CCC     0.00000000 0.99786619  
Default 0.00000

<center>Multiplying $\Phi$ by itself 1000 times is converging to: </center>

$$ \Phi^{1000} \approx \begin{bmatrix}
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\
\end{bmatrix}
$$

<center>Implying that over the very long-term, according to this markov process, the probability of default is $\approx$ 100% no matter the starting state.</center>

$$ \mathbf{p} = \begin{bmatrix} 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \end{bmatrix}$$

# 2. Portfolio Optimization

$$u_0 = \mathbf{h}^T\mathbf{\mu} \tag{1}$$

$$ \mathbf{h}^T\mathbf{1} = 1.00 \tag{2}$$

$$ U(\mathbf{h}) = \mathbf{h}^T\mathbf{\mu} - \frac{\gamma}{2}\mathbf{h}^T\Sigma\mathbf{h} \tag{3}$$

$$A = \mathbf{1}^T\Sigma^{-1}\mathbf{1} \;\;\;\;\;\;\;B=\mathbf{1}^T\Sigma^{-1}\mu\;\;\;\;\;\;C = \mathbf{\mu}^T\Sigma^{-1}\mathbf{\mu} \;\;\;\;\;\;\; \Delta = AC - B^2 > 0 \;\;\;\;\;\;\; \mathbf{w_A} = \frac{1}{A}\Sigma^{-1}\mathbf{1} \;\;\;\;\;\;\; \mathbf{w_B} = \frac{1}{B}\Sigma^{-1}\mathbf{\mu}$$

><b>(a) Derive a formula for the optimal portfolio, $\mathbf{h}$, satisfying (1) and (2) and maximizing (3). Express the formula as a function of A, B, C, $\Delta$, $\mu_0$, $w_A$ and $w_B$</b>

$$L(\mathbf{h}, \lambda_1, \lambda_2) = \mathbf{h}^T\mu - \frac{\gamma}{2}\mathbf{h}^T\Sigma\mathbf{h} + \lambda_1(\mathbf{h}^T\mathbf{1} - 1) + \lambda_2(\mathbf{h}^T\mu - \mu_0)$$

$$\frac{\partial{L}}{\partial{\mathbf{h}}} = \mu - \gamma\Sigma\mathbf{h} + \lambda_1\mathbf{1} + \lambda_2\mathbf{\mu}$$

$$ \mu - \gamma\Sigma\mathbf{h} + \lambda_1\mathbf{1} + \lambda_2\mathbf{\mu} = \mathbf{0}$$

$$- \gamma\Sigma\mathbf{h} = -\mu -\lambda_1\mathbf{1} - \lambda_2\mathbf{\mu} $$

$$\Sigma\mathbf{h} = \frac{1}{\gamma}\mu + \frac{\lambda_1}{\gamma}\mathbf{1} + \frac{\lambda_2}{\gamma}\mathbf{\mu} $$

$$\Sigma^{-1}\Sigma\mathbf{h} = \frac{1}{\gamma}\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\Sigma^{-1}\mathbf{\mu} $$

$$\mathbf{h} = \frac{1}{\gamma}\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\Sigma^{-1}\mathbf{\mu} \tag{4}$$

<center>Will solve the two below equations for $\lambda_1$ and $\lambda_2$</center>

$$\mathbf{1}^T\mathbf{h} = \frac{1}{\gamma}\mathbf{1}^T\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\mathbf{1}^T\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\mathbf{1}^T\Sigma^{-1}\mathbf{\mu} = 1 \;\;\;\;\;\;\;\;\;\; \mathbf{\mu}^T\mathbf{h} = \frac{1}{\gamma}\mathbf{\mu}^T\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\mathbf{\mu}^T\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\mathbf{\mu}^T\Sigma^{-1}\mathbf{\mu} = \mu_{0}$$ 

$$\frac{1}{\gamma}\mathbf{1}^T\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\mathbf{1}^T\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\mathbf{1}^T\Sigma^{-1}\mathbf{\mu} = 1 $$

$$B + \lambda_1A + \lambda_2B = \gamma $$

$$\lambda_1A  = \gamma - B - \lambda_2B $$

$$\lambda_1  = \frac{\gamma - B(1 + \lambda_2)}{A} \tag{5}$$

<center>Plugging this result into the second equation above</center>

$$\frac{1}{\gamma}\mathbf{\mu}^T\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\mathbf{\mu}^T\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\mathbf{\mu}^T\Sigma^{-1}\mathbf{\mu} = \mu_{0}$$ 

$$\frac{1}{\gamma}C + \bigg(\frac{\gamma - B(1 + \lambda_2)}{A}\bigg)\bigg(\frac{\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}}{\gamma}\bigg) + \frac{\lambda_2}{\gamma}C = \mu_{0}$$

$$AC + (\gamma - B - \lambda_2B)(\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}) + \lambda_2AC = \mu_{0}A\gamma$$

$$(\gamma - B)(\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}) - \lambda_2B\mathbf{\mu}^T\Sigma^{-1}\mathbf{1} + \lambda_2AC = \mu_{0}A\gamma - AC$$

$$\lambda_2B\mathbf{\mu}^T\Sigma^{-1}\mathbf{1} = -\mu_{0}A\gamma + AC + (\gamma - B)(\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}) + \lambda_2AC$$

$$\lambda_2 = \frac{-\mu_{0}A\gamma}{B\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}} + \frac{AC}{B\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}} + \frac{(\gamma - B)(\mathbf{\mu}^T\Sigma^{-1}\mathbf{1})}{B\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}} + \frac{\lambda_2AC}{B\mathbf{\mu}^T\Sigma^{-1}\mathbf{1}}$$

<center>Note: since $\Sigma$ is symmetric, $\Sigma^{-1}$ is also symmetric $\implies \mu^T\Sigma^{-1}\mathbf{1} = \mathbf{1}^T\Sigma^{-1}\mathbf{\mu} = B$

$$\lambda_2 = \frac{-\mu_{0}A\gamma}{B^2} + \frac{AC}{B^2} + \frac{(\gamma - B)}{B} + \frac{\lambda_2AC}{B^2}$$

$$\lambda_2 = \frac{AC + \gamma B - B^2 + \lambda_2AC -\mu_{0}A\gamma}{B^2}$$

$$\lambda_2 - \lambda_2\frac{AC}{B^2}= \frac{AC + \gamma B - B^2  -\mu_{0}A\gamma}{B^2}$$

$$\lambda_2 (1- \frac{AC}{B^2})= \frac{AC + \gamma B - B^2  -\mu_{0}A\gamma}{B^2}$$

$$\lambda_2 (B^2- AC)= AC + \gamma B - B^2 -\mu_{0}A\gamma$$

$$\lambda_2 = \frac{\Delta + \gamma B + -\mu_{0}A\gamma}{-\Delta}$$

<center>Plugging $\lambda_2$ into (5) above.</center>

$$\lambda_1  = \frac{1}{A}(\gamma - B(1 + \lambda_2)) \tag{5}$$

$$\lambda_1  = \frac{1}{A}\bigg[\gamma - B\bigg(1 +  \frac{\Delta + \gamma B -\mu_{0}A\gamma}{-\Delta}\bigg)\bigg]$$

$$-A\Delta\lambda_1  = -\Delta\gamma + B(-\Delta +  \Delta + \gamma B -\mu_{0}A\gamma)$$

$$-A\Delta\lambda_1  = -\Delta\gamma + B(\gamma B -\mu_{0}A\gamma)$$

$$-A\Delta\lambda_1  = -\Delta\gamma + \gamma B^2 - \mu_{0}\gamma AB$$

$$\lambda_1  = \frac{\Delta\gamma - \gamma B^2 + \mu_{0}\gamma AB}{A\Delta}$$

<center>Now plugging $\lambda_1$ and $\lambda_2$ back into (4)</center>

$$\mathbf{h} = \frac{1}{\gamma}\Sigma^{-1}\mu + \frac{\lambda_1}{\gamma}\Sigma^{-1}\mathbf{1} + \frac{\lambda_2}{\gamma}\Sigma^{-1}\mathbf{\mu} \tag{4}$$

$$\gamma\mathbf{h} = \Sigma^{-1}\mu + \lambda_1\Sigma^{-1}\mathbf{1} + \lambda_2\Sigma^{-1}\mathbf{\mu}$$

$$\gamma\mathbf{h} = \Sigma^{-1}\mu + \frac{\Delta\gamma - \gamma B^2 + \mu_{0}\gamma AB}{\Delta}\mathbf{w}_A + \frac{-\Delta - \gamma B + \mu_{0}A\gamma}{\Delta}\Sigma^{-1}\mathbf{\mu}$$

$$\Delta\gamma\mathbf{h} = \Delta\Sigma^{-1}\mu + (\Delta\gamma - \gamma B^2 + \mu_{0}\gamma AB)\mathbf{w}_A + (-\Delta - \gamma B + \mu_{0}A\gamma)\Sigma^{-1}\mathbf{\mu}$$

$$\frac{\Delta\gamma}{B}\mathbf{h} = \Delta\mathbf{w}_B + \frac{\gamma\mathbf{w}_A}{B}(\Delta - B^2 + \mu_{0} AB) + \mathbf{w}_B(-\Delta - \gamma B + \mu_{0}A\gamma)$$

$$\frac{\Delta\gamma}{B}\mathbf{h} = \frac{\gamma\mathbf{w}_A}{B}(\Delta - B^2 + \mu_{0} AB) + \gamma\mathbf{w}_B(B + \mu_{0}A)$$

$$\Delta\mathbf{h} = \mathbf{w}_A(AC - 2B^2 + \mu_{0} AB) + \mathbf{w}_B(B^2 + \mu_{0}AB)$$

$$\mathbf{h} = \frac{\mathbf{w}_A(AC - 2B^2 + \mu_{0} AB) + \mathbf{w}_B(B^2 + \mu_{0}AB)}{\Delta}$$

><b>(b) Use your results from (a) to write $\sigma^2_h$ as a quadratic function of $\mu_0$ using only the parameters A, B, C and $\Delta$. This function is known as the minimum variance frontier.</center>

><b>(c) Now assume the investor actually chooses $\mu_0$ to maximize (6), given risk aversion coefficient $\gamma$. Use your previous results to write the optimal $\mu_0$ as a function of the previous parameters and $\gamma$. </b>

# 3. Portfolio Optimization with a Risk-Free Asset

><b>(a) Show that there is a so-called one-fund separation in the market in this case, in that all investors, regardless of their $\gamma$, will hold the same stock market portfolio:</b>

$$\mathbf{w} = \frac{1}{B - AR}\Sigma^{-1}(\mathbf{\mu} - R\mathbf{1})$$

$$ U(\mathbf{h}) = R + \mathbf{h}^T(\mu - R\mathbf{1}) - \frac{\gamma}{2}\mathbf{h}^T\Sigma\mathbf{h} $$

$$ \frac{\partial{U}}{\partial{h}} = \mu - R\mathbf{1} - \gamma\Sigma\mathbf{h}  = 0$$

$$\gamma\Sigma\mathbf{h} = \mu - R\mathbf{1} $$